Allow plugins to declare new arguments

lefessan opened the pull request:
Now that ocamlc has plugins, such plugins may want to declare their own arguments.

lefessan committed:
Allow plugins to declare new arguments with Clflags.add_arguments

dsheets commented:
Is there a plan to prevent plugin and compiler arguments from interfering? Is there a plan to prevent plugin arguments from two unrelated plugins from interfering with each other?

lefessan commented:
I imagine it's more the problem of plugins, not of the compiler. If a plugin wants, it can remove existing arguments, or check if an argument already exists before adding it.

dsheets commented:
Respectfully, that plan does not appear to anticipate the success of this feature.

ghost commented:
In ppx_core we have some code to detect clashes of globally named elements and print a helpful error messages pointing to the place where the two things where registered. For instance:

```
Warning: code transformation custom_printf registered twice.
  - first time was at ppx_custom_printf.ml:337
  - second time is at ppx_fail.ml:19
```

While these kind of errors are indeed the fault of the plugin, having this kind of safety check helps debugging a lot. The code to get the location of the caller is [here](https://github.com/janestreet/ppx_core/blob/master/src/caller_id.ml) and is used this way:

``` ocaml
let caller_id = Caller_id.get ~skip:[__FILE__] in
...
```

Maybe we could do something similar for this feature? We would need to export an `add_arg` function instead of the reference directly.

lefessan commented:
I am not saying we should not provide such helpers, just that this PR only provides the basic mechanism that is needed by the plugins, while another PR later can provide the helpers.

avsm commented:
I think @diml's point is that the raw ref should not be exposed directly to plugins, and that they should go through the helper functions which guarantee that any extra plugin CLI arguments will be checked for clashes.

lefessan commented:
The `ref` cannot not be exposed directly, because it is used (as a `ref`) by `Arg.parse_dynamic`. So, it is possible to provide an helper function, but not to prevent plugins from accessing directly the arguments (unless we wrap `Arg.parse_dynamic` too). I think it is a lot of code for a very small problem, that did not even appear yet.

ghost commented:
Wrapping `Arg.parse_dynamic` is a one-liner. Even if we don't implement the check now, it will make the API future-proof

lefessan commented:
Future proof ? How can we forcast what plugin writers will want to do, and won't be able to do just because it was decided to abstract this value. Why not do the same for all references in `Clflags`, and write setters and getters for all of them, checking that they are being consistently modified with whatever invariants we decide for them ? I think we are adding limitations without any reason. 

Moreover, what should happen if two plugins declare the same option ? Shall we raise an error, preventing the use of the two plugins together even if the option is not actually used ? Shall we instead declare an option that will raise the error only if it is used ? Shall we have different cases if the existing options was declared by the compiler or by another plugin ? Shall we see option handlers as hooks, that could be combined by different plugins (i.e. an option could be declared by several plugins (and the compiler), that would just add some code to be executed when the option is called) ? Sounds like a lot of cases to handle to make the API future-proof.

dsheets commented:
How about namespacing plugin arguments under something like `--x-[pluginshortname]-[pluginarg]`?

ghost commented:
I like the namespacing idea. Another advantage of adopting a uniform naming convention is that when someone report an error and paste a compilation command, we'll be able to tell immediately which arguments are for plugins and which ones are for the compiler itself

lefessan commented:
I implemented a check for already defined arguments, using the function `Clflags.add_arguments`. As a consequence, it discovered that `-keep-docs` was duplicated in the compiler. I added the PR to Milestone 4.04, as the bug fix might make this (very short) PR relevant for the upcoming release.

damiendoligez commented on utils/clflags.ml:
```diff
@@ -358,3 +358,18 @@ let parse_color_setting = function
+    try
+      let loc2 = Misc.StringMap.find arg_name !arg_names in
+      Printf.eprintf
+        "Warning: plugin argument %s is already defined:\n" arg_name;
```
The warning is not in emacs format, so it will get lost in the noise of the build. I'm not sure what can be done about this.

lefessan replied:
Maybe I could print the warning as located at the first line of the plugin ?

damiendoligez replied:
Do you still have the file name of the plugin's source at this point??

alainfrisch replied:
I guess this can be extracted from the callstack at registration time.  That said, it might be more user friendly to have each plugin provide a symbolic name and use that name in the error message.

damiendoligez replied:
> I guess this can be extracted from the callstack at registration time.

Please refrain from using java-style reflection tricks, thank you very much. I guess this problem can wait.

damiendoligez replied:
In fact you have `loc` and `loc2` available here. Why not make it a proper warning and use `Location.print_warning` twice?

Kakadu commented:
@lefessan Do you have examples of helloworld plugins somewhere?

damiendoligez commented:
@lefessan what do you think of namespacing? I like the idea (although I'd drop the `x-` part).

bobzhang commented:
FYI, we do extend the ocaml compiler with more flags in bucklescript, all flags are prefixed with `-bs` to prevent clash with the future release of ocaml compiler.
Another issue I noticed is that not too many ppx used qualified attributes which will cause more problems in the future

Kakadu commented:
Folks, how far away are we from full support of compiler plugins which will, for example, allow to implement additional rewriting of typed tree?

damiendoligez commented:
We still haven't completely agreed on the design, and I just learned that @lefessan will be unavailable for some time. Since this feature is not important enough to delay the release, I'm removing the 4.04 milestone.

Kakadu commented:
@damiendoligez Can you briefly describe current status? I kind of want this feature, so I need to decide will I put some effort and time into it in the nearest future.

damiendoligez commented:
I certainly expect the feature to be included in 4.05, and probably even in 4.04.1 if there is a 4.04.1.
AFAICT we just need to agree on the namespacing thing.

Concerning full support for compiler plugins, I think we're very close but you'll have to wait for Fabrice to come back to get a full answer.

lefessan commented:
Here is an updated version. I am not fond of namespacing on arguments, as it may prevent us from adding simple arguments. I would prefer plugin developers to do it by themselves if they want their plugin to follow such a discipline. Anyway, I think we can merge this PR, and wait for a proposal for namespaces, and decide at that point what to do next.

lefessan commented:
Since there was no discussion anymore since the last version, I think we reached a satisfying state.
Adding namespaces would be trivial, if really necessary (add an optional namespace argument to `Clflags.add_arguments`). Thus merged.
